eZ/Publish/Core/Persistence/Legacy/User/Role/Gateway/DoctrineDatabase.php

```diff
@@ -334,6 +334,73 @@ public function loadRoleAssignmentsByRoleId(int $roleId): array
         return $statement->fetchAll(FetchMode::ASSOCIATIVE);
     }
 
+    /**
+     * @throws \Doctrine\DBAL\Driver\Exception
+     * @throws \Doctrine\DBAL\Exception
+     */
+    public function loadRoleAssignmentsByRoleIdWithOffsetAndLimit(int $roleId, int $offset, ?int $limit): array
+    {
+        $query = $this->connection->createQueryBuilder();
+        $query->select(
+            'user_role.id',
+            'user_role.contentobject_id',
+            'user_role.limit_identifier',
+            'user_role.limit_value',
+            'user_role.role_id'
+        )->from(
+            self::USER_ROLE_TABLE,
+            'user_role'
+        )->innerJoin(
+            'user_role',
+            self::CONTENT_OBJECT_TABLE,
+            'content_object',
+            'user_role.contentobject_id = content_object.id'
+        )->where(
+            $query->expr()->eq(
+                'role_id',
+                $query->createPositionalParameter($roleId, ParameterType::INTEGER)
+            )
+        )
+        ->setFirstResult($offset);
+
+        if ($limit !== null) {
+            $query->setMaxResults($limit);
+        }
+
+        return $query
+            ->execute()
+            ->fetchAllAssociative();
+    }
+
+    /**
+     * @throws \Doctrine\DBAL\Driver\Exception
+     * @throws \Doctrine\DBAL\Exception
+     */
+    public function countRoleAssignments(int $roleId): int
+    {
+        $query = $this->connection->createQueryBuilder();
+
+        $query
+            ->select($this->connection->getDatabasePlatform()->getCountExpression('user_role.id'))
+            ->innerJoin(
+                'user_role',
+                self::CONTENT_OBJECT_TABLE,
+                'content_object',
+                'user_role.contentobject_id = content_object.id'
+            )
+            ->from(
+                self::USER_ROLE_TABLE,
+                'user_role'
+            )->where(
+                $query->expr()->eq(
+                    'role_id',
+                    $query->createPositionalParameter($roleId, ParameterType::INTEGER)
+                )
+            );
```

Comment:
This is almost ok. Now notice that you have 2x almost the same code.

If you extract the common part as:

```php
    /**
     * @param array<string> $columns
     */
    private function buildLoadRoleAssignmentsQuery(array $columns, int $roleId): QueryBuilder
    {
        $query = $this->connection->createQueryBuilder();
        $query
            ->select(...$columns)
            ->from(self::USER_ROLE_TABLE, 'user_role')
            ->innerJoin(
                'user_role',
                self::CONTENT_OBJECT_TABLE,
                'content_object',
                'user_role.contentobject_id = content_object.id'
            )->where(
                $query->expr()->eq(
                    'role_id',
                    $query->createPositionalParameter($roleId, ParameterType::INTEGER)
                )
            );

        return $query;
    }
```

then `countRoleAssignments` could have the form of:
```php
        $query = $this->buildLoadRoleAssignmentsQuery(
            [$this->connection->getDatabasePlatform()->getCountExpression('user_role.id')],
            $roleId
        );

        return (int)$query->execute()->fetchOne();
```

and `loadRoleAssignmentsByRoleIdWithOffsetAndLimit`:
```php
        $query = $this
            ->buildLoadRoleAssignmentsQuery(
                [
                    'user_role.id',
                    'user_role.contentobject_id',
                    'user_role.limit_identifier',
                    'user_role.limit_value',
                    'user_role.role_id',
                ],
                $roleId
            )
            ->setFirstResult($offset);

        if ($limit !== null) {
            $query->setMaxResults($limit);
        }

        return $query
            ->execute()
            ->fetchAllAssociative();
```

It's crucial that pagination and count for pagination re-uses the same query to ensure the stability of the solution :-)

Reply:
@alongosz good idea as queries are matching right now, resolved in https://github.com/ezsystems/ezplatform-kernel/pull/373/commits/156d25b81696240e50b7ddb77ff1e0b38b591c42